### PR TITLE
Message Bus

### PR DESCRIPTION
**Pull request by ASKozienko**

### CODE REVIEW

File: pkg/mq-amqp-ext/Client/AmqpDriver.php [hunk 1]
```diff
@@ -50,6 +58,56 @@ public function __construct(AmqpContext $context, Config $config)
 
     /**
      * {@inheritdoc}
+     */
+    public function sendToRouter(Message $message)
+    {
+        $topic = $this->createRouterTopic();
```
Comment (makasim) [inline pkg/mq-amqp-ext/Client/AmqpDriver.php]: we have to check that the topic was set.

File: pkg/mq-amqp-ext/Client/AmqpDriver.php [hunk 2]
```diff
@@ -50,6 +58,56 @@ public function __construct(AmqpContext $context, Config $config)
...
+
+        $this->context->createProducer()->send($topic, $transportMessage);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function sendToProcessor(Message $message)
+    {
+        if (false == $queueName = $message->getProperty(Config::PARAMETER_PROCESSOR_QUEUE_NAME)) {
```
Comment (makasim) [inline pkg/mq-amqp-ext/Client/AmqpDriver.php]: also we have to check that a processor name is set.

File: pkg/mq-amqp-ext/Symfony/AmqpTransportFactory.php [hunk 3]
```diff
@@ -98,6 +99,24 @@ public function createService(ContainerBuilder $container, array $config)
     /**
      * {@inheritdoc}
      */
+    public function createDriver(ContainerBuilder $container, array $config)
```
Comment (makasim) [inline pkg/mq-amqp-ext/Symfony/AmqpTransportFactory.php]: let's rename createService to createContext

File: pkg/mq/Client/DelegateMessageProcessor.php [hunk 4]
```diff
@@ -10,14 +10,21 @@ class DelegateMessageProcessor implements MessageProcessorInterface
...
 
     /**
      * @param MessageProcessorRegistryInterface $registry
+     * @param Config                            $config
      */
-    public function __construct(MessageProcessorRegistryInterface $registry)
+    public function __construct(MessageProcessorRegistryInterface $registry, Config $config)
     {
         $this->registry = $registry;
+        $this->config = $config;
```
Comment (makasim) [inline pkg/mq/Client/DelegateMessageProcessor.php]: seems like it is not used

File: pkg/mq/Client/NullDriver.php [hunk 5]
```diff
@@ -113,4 +113,19 @@ public function send(Queue $queue, Message $message)
...
+    }
+
+    public function sendToProcessor(Message $message)
+    {
+        // TODO: Implement sendToProcessor() method.
+    }
+
+    public function setupBroker()
+    {
+        // TODO: Implement setupBroker() method.
```
Comment (makasim) [inline pkg/mq/Client/NullDriver.php]: remove comments.

File: pkg/mq/Client/NullDriver.php [hunk 6]
```diff
@@ -113,4 +113,19 @@ public function send(Queue $queue, Message $message)
...
+    {
+        // TODO: Implement sendToRouter() method.
+    }
+
+    public function sendToProcessor(Message $message)
+    {
+        // TODO: Implement sendToProcessor() method.
+    }
+
+    public function setupBroker()
```
Comment (makasim) [inline pkg/mq/Client/NullDriver.php]: add doc blocks

File: pkg/mq/Client/RouterProcessor.php [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            ));
+        }
+
+        if (array_key_exists($topicName, $this->routes)) {
+            foreach ($this->routes[$topicName] as $route) {
+                $processorMessage = clone $message;
+                $processorMessage->setProperty(Config::PARAMETER_PROCESSOR_NAME, $route[0]);
+                $processorMessage->setProperty(Config::PARAMETER_PROCESSOR_QUEUE_NAME, $route[1]);
+
+                $this->driver->sendToProcessor($this->driver->createClientMessage($processorMessage));
```
Comment (makasim) [inline pkg/mq/Client/RouterProcessor.php]: maybe we can unset delay here, since we are creating a client's message object here in any case

**Comment by makasim** [comment]
create queues command has to be updated too - https://github.com/formapro/message-queue-dev/blob/master/pkg/mq/Client/CreateQueuesCommand.php

File: pkg/mq/Client/Config.php [hunk 8]
```diff
@@ -5,8 +5,9 @@ class Config
 {
     const PARAMETER_TOPIC_NAME = 'formapro.message_queue.client.topic_name';
```
Comment (makasim) [inline pkg/mq/Client/Config.php]: What about rename this to `formapro.topic_name`

File: pkg/mq/Client/ConsumptionExtension/DelayRedeliveredMessageExtension.php [hunk 9]
```diff
@@ -12,7 +11,7 @@ class DelayRedeliveredMessageExtension implements ExtensionInterface
 {
     use EmptyExtensionTrait;
 
-    const PROPERTY_REDELIVER_COUNT = 'fp-redeliver-count';
+    const PROPERTY_REDELIVER_COUNT = 'formapro-redeliver-count';
```
Comment (makasim) [inline pkg/mq/Client/ConsumptionExtension/DelayRedeliveredMessageExtension.php]: Rename to `formapro.redelivery_count`

File: pkg/mq/Client/Config.php [hunk 10]
```diff
@@ -5,8 +5,9 @@ class Config
...
     const PARAMETER_PROCESSOR_NAME = 'formapro.message_queue.client.processor_name';
-    const PARAMETER_QUEUE_NAME = 'formapro.message_queue.client.queue_name';
-    const DEFAULT_QUEUE_NAME = 'default';
+    const PARAMETER_PROCESSOR_QUEUE_NAME = 'formapro.message_queue.client.processor_queue_name';
+    const SHARED_ROUTER_PROCESSOR_NAME = '__router__';
```
Comment (ASKozienko) [inline pkg/mq/Client/Config.php]: remove

Comment (makasim) [inline pkg/mq/Client/Config.php]: why?

File: pkg/mq/Client/ConsumptionExtension/SetupBrokerExtension.php [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function onStart(Context $context)
+    {
+        if (false == $this->isDone) {
+            $this->isDone = true;
+            $this->driver->setupBroker();
```
Comment (makasim) [inline pkg/mq/Client/ConsumptionExtension/SetupBrokerExtension.php]: we are not able to use this logger at this stage (this a note to me)

**Pull request merged**